File: typescript/src/patterns/observer/OrderSubject.ts
import type { Subject } from './Subject.js';
import type { Observer } from './Observer.js';

export type OrderStatus = 'created' | 'paid' | 'shipped';

export class Order implements Subject<OrderStatus> {
  private observers = new Set<Observer<OrderStatus>>();
  private _status: OrderStatus = 'created';

  attach(o: Observer<OrderStatus>): void { this.observers.add(o); }
  detach(o: Observer<OrderStatus>): void { this.observers.delete(o); }
  notify(value: OrderStatus): void { this.observers.forEach(o => o.update(value)); }

  get status(): OrderStatus { return this._status; }
  set status(s: OrderStatus) { this._status = s; this.notify(this._status); }
}

File: python/src/patterns/observer/order_subject.py
from typing import Literal
from .subject import Subject

OrderStatus = Literal['created','paid','shipped']

class Order(Subject[OrderStatus]):
    def __init__(self) -> None:
        super().__init__()
        self._status: OrderStatus = 'created'
    @property
    def status(self) -> OrderStatus:
        return self._status
    @status.setter
    def status(self, s: OrderStatus) -> None:
        self._status = s
        self.notify(self._status)
